File: backend/app/services/quota_service.py
"""
Quota service for managing monthly token limits.

Handles:
- Checking remaining quota before requests
- Atomic token deduction after OpenAI responses
- Monthly row creation (row-per-month pattern)
- Race condition prevention with database locking

Design decisions:
- Row-per-month: Simple monthly resets without background jobs
- SELECT FOR UPDATE: Prevents double-spending with row-level locks
- Optimistic pre-check + pessimistic deduction: Fast reads, safe writes
- Database as source of truth: Survives restarts, works multi-instance
"""
from datetime import datetime
from typing import Tuple, Optional
from sqlalchemy.orm import Session
from sqlalchemy import and_
from sqlalchemy.exc import IntegrityError
import logging

from app.models.token_usage import TokenUsage
from app.models.user import User, TIER_TOKEN_LIMITS, SubscriptionTier

logger = logging.getLogger(__name__)

# Maximum tokens per request (server-side clamp)
MAX_TOKENS_PER_REQUEST = 512


class QuotaExceededError(Exception):
    """Raised when user has exceeded their monthly token quota."""
    def __init__(self, used: int, limit: int, reset_date: str):
        self.used = used
        self.limit = limit
        self.reset_date = reset_date
        super().__init__(f"Monthly token quota exceeded: {used}/{limit}. Resets {reset_date}")


class InsufficientQuotaError(Exception):
    """Raised when user doesn't have enough quota for the estimated request."""
    def __init__(self, remaining: int, required: int):
        self.remaining = remaining
        self.required = required
        super().__init__(f"Insufficient quota: {remaining} remaining, ~{required} required")


def get_current_period() -> Tuple[int, int]:
    """Get current year and month for quota period."""
    now = datetime.utcnow()
    return now.year, now.month


def get_next_reset_date() -> str:
    """Get the date string when quota resets (1st of next month)."""
    now = datetime.utcnow()
    if now.month == 12:
        return f"{now.year + 1}-01-01"
    return f"{now.year}-{now.month + 1:02d}-01"


def get_or_create_usage_record(db: Session, user_id: int) -> TokenUsage:
    """
    Get or create the usage record for current month.
    
    Uses INSERT ... ON CONFLICT pattern to handle race conditions
    when multiple requests try to create the first record of the month.
    """
    year, month = get_current_period()
    
    # Try to get existing record first (fast path)
    usage = db.query(TokenUsage).filter(
        and_(
            TokenUsage.user_id == user_id,
            TokenUsage.year == year,
            TokenUsage.month == month
        )
    ).first()
    
    if usage:
        return usage
    
    # Create new record for this month
    try:
        usage = TokenUsage(
            user_id=user_id,
            year=year,
            month=month,
            tokens_used=0,
            request_count=0
        )
        db.add(usage)
        db.commit()
        db.refresh(usage)
        return usage
    except IntegrityError:
        # Race condition: another request created it first
        db.rollback()
        return db.query(TokenUsage).filter(
            and_(
                TokenUsage.user_id == user_id,
                TokenUsage.year == year,
                TokenUsage.month == month
            )
        ).first()


def check_quota(db: Session, user: User) -> Tuple[int, int, int]:
    """
    Check user's remaining quota without modifying it.
    
    This is a fast, read-only check for pre-flight validation.
    The actual deduction happens atomically after the OpenAI call.
    
    Args:
        db: Database session
        user: User object with subscription tier
        
    Returns:
        Tuple of (remaining_tokens, used_tokens, limit_tokens)
        
    Raises:
        QuotaExceededError: If quota is already exceeded
    """
    usage = get_or_create_usage_record(db, user.id)
    limit = user.monthly_token_limit
    used = usage.tokens_used
    remaining = max(0, limit - used)
    
    if remaining == 0:
        raise QuotaExceededError(used, limit, get_next_reset_date())
    
    return remaining, used, limit


def check_quota_sufficient(db: Session, user: User, estimated_tokens: int) -> bool:
    """
    Check if user has enough quota for an estimated request.
    
    Args:
        db: Database session  
        user: User object
        estimated_tokens: Estimated tokens for the request
        
    Returns:
        True if sufficient quota, raises exception otherwise
        
    Raises:
        QuotaExceededError: If quota is fully exhausted
        InsufficientQuotaError: If remaining < estimated tokens
    """
    remaining, used, limit = check_quota(db, user)
    
    # Allow request if user has ANY remaining quota
    # We can't know exact usage until after the call
    # The hard limit is enforced by the atomic deduction
    if remaining < estimated_tokens:
        # Warning: user is close to limit, but allow if they have anything left
        logger.warning(
            f"User {user.id} has {remaining} tokens remaining, "
            f"request may use up to {estimated_tokens}"
        )
    
    return True


def deduct_tokens_atomic(db: Session, user_id: int, tokens_used: int) -> TokenUsage:
    """
    Atomically deduct tokens from user's monthly quota.
    
    Uses SELECT FOR UPDATE to prevent race conditions where multiple
    concurrent requests could overspend the quota.
    
    This is called AFTER receiving OpenAI's response with actual token count.
    Even if this causes the user to exceed their limit, we deduct it
    (they already used the tokens). Future requests will be blocked.
    
    Args:
        db: Database session
        user_id: User's ID
        tokens_used: Actual tokens used (from OpenAI's usage.total_tokens)
        
    Returns:
        Updated TokenUsage record
    """
    year, month = get_current_period()
    
    # SELECT FOR UPDATE locks the row to prevent concurrent updates
    # This is the critical section that prevents double-spending
    usage = db.query(TokenUsage).filter(
        and_(
            TokenUsage.user_id == user_id,
            TokenUsage.year == year,
            TokenUsage.month == month
        )
    ).with_for_update().first()
    
    if not usage:
        # Should not happen if check_quota was called first, but handle it
        usage = get_or_create_usage_record(db, user_id)
        # Re-acquire with lock
        usage = db.query(TokenUsage).filter(
            and_(
                TokenUsage.user_id == user_id,
                TokenUsage.year == year,
                TokenUsage.month == month
            )
        ).with_for_update().first()
    
    # Update atomically
    usage.tokens_used += tokens_used
    usage.request_count += 1
    
    db.commit()
    db.refresh(usage)
    
    logger.info(
        f"User {user_id} used {tokens_used} tokens. "
        f"Monthly total: {usage.tokens_used}"
    )
    
    return usage


def get_usage_stats(db: Session, user: User) -> dict:
    """
    Get user's current usage statistics for display.
    
    Returns:
        Dict with usage stats including remaining, used, limit, etc.
    """
    usage = get_or_create_usage_record(db, user.id)
    limit = user.monthly_token_limit
    used = usage.tokens_used
    remaining = max(0, limit - used)
    
    return {
        "tokens_used": used,
        "tokens_remaining": remaining,
        "tokens_limit": limit,
        "request_count": usage.request_count,
        "subscription_tier": user.subscription_tier,
        "period": f"{usage.year}-{usage.month:02d}",
        "resets_at": get_next_reset_date(),
        "usage_percentage": round((used / limit) * 100, 1) if limit > 0 else 100
    }


def clamp_max_tokens(requested_max_tokens: Optional[int]) -> int:
    """
    Clamp the max_tokens parameter to server-side limit.
    
    Never trust client-provided max_tokens - always enforce server limit.
    """
    if requested_max_tokens is None:
        return MAX_TOKENS_PER_REQUEST
    return min(requested_max_tokens, MAX_TOKENS_PER_REQUEST)

File: backend-node/src/services/quotaService.js
/**
 * Quota service for managing monthly token limits.
 *
 * Handles:
 * - Checking remaining quota before requests
 * - Atomic token deduction after OpenAI responses
 * - Monthly row creation (row-per-month pattern)
 * - Race condition prevention with database locking
 */
const { Op } = require('sequelize');
const { sequelize } = require('../database');
const { TokenUsage } = require('../models');
const { TIER_TOKEN_LIMITS, SubscriptionTier } = require('../models');

// Maximum tokens per request (server-side clamp)
const MAX_TOKENS_PER_REQUEST = 512;

/**
 * Error thrown when user has exceeded their monthly token quota.
 */
class QuotaExceededError extends Error {
  constructor(used, limit, resetDate) {
    super(`Monthly token quota exceeded: ${used}/${limit}. Resets ${resetDate}`);
    this.name = 'QuotaExceededError';
    this.used = used;
    this.limit = limit;
    this.resetDate = resetDate;
  }
}

/**
 * Error thrown when user doesn't have enough quota for the estimated request.
 */
class InsufficientQuotaError extends Error {
  constructor(remaining, required) {
    super(`Insufficient quota: ${remaining} remaining, ~${required} required`);
    this.name = 'InsufficientQuotaError';
    this.remaining = remaining;
    this.required = required;
  }
}

/**
 * Get current year and month for quota period.
 * @returns {[number, number]} Tuple of (year, month)
 */
function getCurrentPeriod() {
  const now = new Date();
  return [now.getUTCFullYear(), now.getUTCMonth() + 1];
}

/**
 * Get the date string when quota resets (1st of next month).
 * @returns {string} Reset date string
 */
function getNextResetDate() {
  const now = new Date();
  const year = now.getUTCMonth() === 11 ? now.getUTCFullYear() + 1 : now.getUTCFullYear();
  const month = now.getUTCMonth() === 11 ? 1 : now.getUTCMonth() + 2;
  return `${year}-${month.toString().padStart(2, '0')}-01`;
}

/**
 * Get or create the usage record for current month.
 * @param {number} userId - User ID
 * @param {object} [transaction] - Optional transaction
 * @returns {Promise<TokenUsage>} Usage record
 */
async function getOrCreateUsageRecord(userId, transaction = null) {
  const [year, month] = getCurrentPeriod();

  // Try to get existing record first (fast path)
  let usage = await TokenUsage.findOne({
    where: {
      userId,
      year,
      month,
    },
    transaction,
  });

  if (usage) {
    return usage;
  }

  // Create new record for this month
  try {
    usage = await TokenUsage.create({
      userId,
      year,
      month,
      tokensUsed: 0,
      requestCount: 0,
    }, { transaction });
    return usage;
  } catch (error) {
    // Race condition: another request created it first
    if (error.name === 'SequelizeUniqueConstraintError') {
      return TokenUsage.findOne({
        where: {
          userId,
          year,
          month,
        },
        transaction,
      });
    }
    throw error;
  }
}

/**
 * Check user's remaining quota without modifying it.
 * @param {User} user - User object with subscription tier
 * @returns {Promise<[number, number, number]>} Tuple of (remaining, used, limit)
 * @throws {QuotaExceededError} If quota is already exceeded
 */
async function checkQuota(user) {
  const usage = await getOrCreateUsageRecord(user.id);
  const limit = user.getMonthlyTokenLimit();
  const used = usage.tokensUsed;
  const remaining = Math.max(0, limit - used);

  if (remaining === 0) {
    throw new QuotaExceededError(used, limit, getNextResetDate());
  }

  return [remaining, used, limit];
}

/**
 * Check if user has enough quota for an estimated request.
 * @param {User} user - User object
 * @param {number} estimatedTokens - Estimated tokens for the request
 * @returns {Promise<boolean>} True if sufficient quota
 * @throws {QuotaExceededError} If quota is fully exhausted
 * @throws {InsufficientQuotaError} If not enough quota for request
 */
async function checkQuotaSufficient(user, estimatedTokens) {
  const [remaining, used, limit] = await checkQuota(user);

  if (remaining < estimatedTokens) {
    throw new InsufficientQuotaError(remaining, estimatedTokens);
  }

  return true;
}

/**
 * Atomically deduct tokens after OpenAI response.
 * Uses transaction with row-level locking to prevent race conditions.
 * @param {number} userId - User ID
 * @param {number} tokensUsed - Actual tokens used from OpenAI response
 * @returns {Promise<TokenUsage>} Updated usage record
 */
async function deductTokensAtomic(userId, tokensUsed) {
  return sequelize.transaction(async (transaction) => {
    const [year, month] = getCurrentPeriod();

    // Lock the row for update
    let usage = await TokenUsage.findOne({
      where: {
        userId,
        year,
        month,
      },
      lock: transaction.LOCK.UPDATE,
      transaction,
    });

    if (!usage) {
      // Create if doesn't exist
      usage = await TokenUsage.create({
        userId,
        year,
        month,
        tokensUsed: 0,
        requestCount: 0,
      }, { transaction });
    }

    // Update atomically
    usage.tokensUsed += tokensUsed;
    usage.requestCount += 1;
    await usage.save({ transaction });

    return usage;
  });
}

/**
 * Clamp max_tokens to server-enforced limit.
 * @param {number|null} requestedMaxTokens - Client-requested max tokens
 * @returns {number} Clamped max tokens value
 */
function clampMaxTokens(requestedMaxTokens) {
  if (requestedMaxTokens === null || requestedMaxTokens === undefined) {
    return MAX_TOKENS_PER_REQUEST;
  }
  return Math.min(Math.max(1, requestedMaxTokens), MAX_TOKENS_PER_REQUEST);
}

/**
 * Get usage statistics for a user.
 * @param {User} user - User object
 * @returns {Promise<object>} Usage statistics
 */
async function getUsageStats(user) {
  const usage = await getOrCreateUsageRecord(user.id);
  const limit = user.getMonthlyTokenLimit();
  const used = usage.tokensUsed;
  const remaining = Math.max(0, limit - used);
  const [year, month] = getCurrentPeriod();

  return {
    tokensUsed: used,
    tokensRemaining: remaining,
    tokensLimit: limit,
    requestCount: usage.requestCount,
    subscriptionTier: user.subscriptionTier,
    period: `${year}-${month.toString().padStart(2, '0')}`,
    resetsAt: getNextResetDate(),
    usagePercentage: Math.round((used / limit) * 100 * 100) / 100,
  };
}

module.exports = {
  MAX_TOKENS_PER_REQUEST,
  QuotaExceededError,
  InsufficientQuotaError,
  getCurrentPeriod,
  getNextResetDate,
  getOrCreateUsageRecord,
  checkQuota,
  checkQuotaSufficient,
  deductTokensAtomic,
  clampMaxTokens,
  getUsageStats,
};
